### PR TITLE
sysusers: make sure to reset the returned value when fget*ent_sane() …

### PR DESCRIPTION
**Pull request by fbuihuu**
…wrapper return false

To indicate that the there're no more entries, these wrappers return false but
leave the passed pointed unmodified.

Fixes: #8721

### CODE REVIEW

File: src/sysusers/sysusers.c [hunk 1]
```diff
@@ -416,6 +416,8 @@ static int write_temporary_passwd(const char *passwd_path, FILE **tmpfile, char
                 }
                 if (r < 0)
                         return r;
+                if (!r)
```
Comment (keszybz) [inline src/sysusers/sysusers.c]: This should be `r != 0`, following the don't-downgrade-ints-to-bool rule.

Comment (fbuihuu) [reply src/sysusers/sysusers.c]: ok but the condition should be `r == 0`

Comment (yuwata) [reply src/sysusers/sysusers.c]: Not fixed yet...

**Comment by fbuihuu** [comment]
damn I missed one... thanks @yuwata for spotting.

**Comment by poettering** [comment]
Hmm, I'd rather see this fixed in fgetpwent_sane() and friends... Our rule is usually to initialize return parameters on sucess, and not touch them on failure. success is usually indicated via a return value of >= 0, while failure is indicated by < 0. Now, fgetpwent_sane() return < 0 on true failure, and == 0 on EOF, which I think is entirely fine, because EOF is not an error, it's a very common case. Hence, I think in this case we should initialize the return parameter properly, i.e. to NULL...

**Comment by keszybz** [comment]
Indeed, that'd be much better.

**Comment by evverx** [comment]
To be honest, I'm not sure the wrappers have made the code easier to understand, so I would say that simply reverting https://github.com/systemd/systemd/commit/100d5f6ee6ec53ecc20160cd4d6656cacf57db4f would probably be better, in a sense.

Also, the word "sane" seems to imply that what is wrapped is kind of insane and appears to have negative connotations. Would it make sense to use something like `putent_errno_wrapper` to make it clear what these functions are supposed to do?

**Comment by fbuihuu** [comment]
@evverx the wrappers have been introduced because the interface provided by the glibc is too error prone: errno must be checked to validate that no errors occured which means that we need to make sure to reset its value every times we call the glibc interfaces, something that can be forgotten/missed easily, see 8c1b45aa9cc0241bdf2c143df493ddf2596ae8b1 for an example.

**Comment by keszybz** [comment]
This seems to be misunderstanding. @poettering's comment was intended to make `fgetpwent_sane` assign NULL to the output argument when it returns 0. Then this wouldn't have to be done in all the callers.

**Comment by fbuihuu** [comment]
ah right, it seems that I will never make it correct :(

**Comment by evverx** [comment]
@fbuihuu thank you. I think I understand now why the wrappers were introduced and I'm inclined to agree that they are useful in the long run. I still don't think that wrapping something and call it "sane" is a good idea, but I'm probably missing some cultural context where it seems to be totally fine.

**Comment by fbuihuu** [comment]
here we go again...

**Comment by keszybz** [comment]
> I still don't think that wrapping something and call it "sane" is a good idea, but I'm probably missing some cultural context where it seems to be totally fine.

It's a passive-agressive way of saying that we don't like the interface ;)

**Comment by fbuihuu** [comment]
No problem for switching to a passive-peace way though :)

**Comment by poettering** [comment]
rawhide CI fails with:

```
In file included from ../src/basic/util.h:47,
                 from ../src/basic/hashmap.h:30,
                 from ../src/shared/bus-util.h:31,
                 from ../src/core/service.c:31:
../src/basic/missing.h:528:17: error: expected identifier before numeric constant
 #define MS_MOVE 8192
                 ^~~~
```

i am puzzled...

**Comment by evverx** [comment]
It seems that the test wouldn't have failed if the branch had contained https://github.com/systemd/systemd/commit/227b8a762fea1458547be2cdf0e6e4aac0079730.

**Comment by fbuihuu** [comment]
done

**Pull request merged**